### PR TITLE
HPCC-18613 Define serialized binary type info format more robustly

### PR DESCRIPTION
**Pull request by richardkchapman**
In order to make sure we can implement future functionality while maintaining
backward compatibility, make some minor tweaks to the way that type info is
serialized in binary form:

  1. Add a "format" marker so we can detect different versions
  2. Define endianness.
  3. Add a hash (so we can cache deserializations efficiently)
  4. Add a size (so we can embed inside other serializations)

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [x] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [x] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Regression suite should cover it.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review

**Comment by AttilaVamos** [comment]
PR is rescheduled.

File: rtl/eclrtl/rtldynfield.cpp [hunk 1]
```diff
@@ -437,14 +440,22 @@ class CRtlFieldTypeBinSerializer
...
+         size32_t pos = out.length();
          if (applyBias)
          {
              IndexBiasTranslator translator(type);
              s.serializeType(translator.queryTranslatedType());
          }
          else
              s.serializeType(type);
+         size.write();
+         hash.write(rtlHash64Data(size.size(), out.toByteArray()+pos, 0));
```
Comment (ghalliday) [inline rtl/eclrtl/rtldynfield.cpp]: Should this also restore the endianness?

File: rtl/eclrtl/rtldynfield.cpp [hunk 2]
```diff
@@ -629,23 +640,46 @@ class CRtlFieldTypeDeserializer : public CInterfaceOf<IRtlFieldTypeDeserializer>
...
+            hash64_t hash;
+            buf.read(hash);
+            size32_t size;
+            buf.read(size);
+#ifdef VALIDATE_TYPEINFO_HASHES
+            hash64_t expected = rtlHash64Data(size, buf.readDirect(0), 0);
+            if (expected != hash)
+                throw MakeStringException(0, "Invalid type info hash in CRtlFieldTypeDeserializer::deserialize");
+#endif
+            while (buf.remaining())
```
Comment (ghalliday) [inline rtl/eclrtl/rtldynfield.cpp]: Should this loop until size has been consumed - e.g., if multiple items have been written to the memory buffer

File: rtl/eclrtl/rtldynfield.cpp (hunk 1 above)
Comment (richardkchapman) [reply rtl/eclrtl/rtldynfield.cpp]: Probably

File: rtl/eclrtl/rtldynfield.cpp (hunk 2 above)
Comment (richardkchapman) [reply rtl/eclrtl/rtldynfield.cpp]: Yes

**Pull request merged**